### PR TITLE
SplashScreen

### PR DESCRIPTION
**Pull request by BQuinteros**
Added splash screen and the functionality to run in another thread the reading of JSON file. Adding the new screen I need to refactor some activities and changed his functionality
ezgif com-video-to-gif (2)

### CODE REVIEW

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt [hunk 1]
```diff
@@ -37,38 +38,33 @@ class CityMainActivity : AppCompatActivity(), CityContract.View {
...
-
-    private fun createCityList(json: String?) {
-        val jsonArray = JSONArray(json)
-        setCityListAdapter(viewModel.createCityList(jsonArray))
-    }
-
-    private fun setCityListAdapter(cities: MutableList<String>) {
-        val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, cities)
+    private fun setCityListAdapter() {
+        val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, listOfCities)
```
Comment (SartiMau) [inline WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt]: from where do you get the listOfCities variable?

File: WheaterApp/app/src/main/java/com/example/weatherapp/viewmodels/SplashScreenViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import com.example.weatherapp.viewmodels.CityMainViewModel.Companion.listOfCity
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.json.JSONArray
+
+class SplashScreenViewModel(
+        private val createCityListUseCase: CreateCityListUseCase
+) : ViewModel(), SplashScreenContract.ViewModel {
+    override val mutableMainState: MutableLiveData<Event<Data<City>>> = MutableLiveData()
```
Comment (lucianoeli) [inline WheaterApp/app/src/main/java/com/example/weatherapp/viewmodels/SplashScreenViewModel.kt]: this one should be private.

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    private fun readJSONFile() {
+        val json = applicationContext.assets.open(FILE_NAME).bufferedReader().use {
+            it.readText()
+        }
+        createCityList(json)
+    }
+
+    override fun nextActivity() {
```
Comment (SartiMau) [inline WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt]: this method is not necessary you can call the start activity from the updateUI method

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        startActivity<CityMainActivity>(null, null)
+    }
+
+    private fun createCityList(json: String) {
+        viewModel.createCityList(json)
```
Comment (SartiMau) [inline WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt:48]: same here, call this from readJSONFile method as this line is used only one time

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    private fun showMessageError() {
+        Toast.makeText(this, MESSAGE, Toast.LENGTH_SHORT).show()
```
Comment (SartiMau) [inline WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt]: same.. you call this only one time

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: WheaterApp/app/src/main/java/com/example/weatherapp/viewmodels/SplashScreenViewModel.kt (hunk 2 above)
Comment (BQuinteros) [reply WheaterApp/app/src/main/java/com/example/weatherapp/viewmodels/SplashScreenViewModel.kt]: Done!

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt (hunk 3 above)
Comment (BQuinteros) [reply WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt]: Done!

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt (hunk 5 above)
Comment (BQuinteros) [reply WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt]: Done!

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt (hunk 4 above)
Comment (BQuinteros) [reply WheaterApp/app/src/main/java/com/example/weatherapp/activities/SplashScreenActivity.kt:48]: Done!

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt (hunk 1 above)
Comment (BQuinteros) [reply WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt]: I'm taking from CityViewModel

**Comment by BQuinteros** [comment]
ezgif com-video-to-gif (3)

**Review by mauroLonghi — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt (hunk 1 above)
Comment (SartiMau) [reply WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt]: fix this based on the just talked

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: WheaterApp/app/src/main/java/com/example/weatherapp/contracts/SplashScreenContract.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import androidx.lifecycle.LiveData
+import com.example.domain.entities.City
+import com.example.weatherapp.utils.Data
+import com.example.weatherapp.utils.Event
+import kotlinx.coroutines.Job
+
+interface SplashScreenContract {
+    interface ViewModel {
+        val mainState: LiveData<Event<Data<City>>>
```
Comment (mauroLonghi) [inline WheaterApp/app/src/main/java/com/example/weatherapp/contracts/SplashScreenContract.kt]: Why you define this variable in the interface?

File: WheaterApp/domain/src/main/java/com/example/domain/usecases/GetJSONData.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.example.domain.usecases
+
+interface GetJSONData {
```
Comment (mauroLonghi) [inline WheaterApp/domain/src/main/java/com/example/domain/usecases/GetJSONData.kt]: Maybe you can group thats interfaces in only one file...not it wrong it a suggestion,

File: WheaterApp/app/src/main/java/com/example/weatherapp/contracts/SplashScreenContract.kt (hunk 6 above)
Comment (BQuinteros) [reply WheaterApp/app/src/main/java/com/example/weatherapp/contracts/SplashScreenContract.kt]: Removed.

File: WheaterApp/domain/src/main/java/com/example/domain/usecases/GetJSONData.kt (hunk 7 above)
Comment (BQuinteros) [reply WheaterApp/domain/src/main/java/com/example/domain/usecases/GetJSONData.kt]: Yes, I prefer to create one file with all the functionality inside

**Review by mauroLonghi — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt [hunk 8]
```diff
@@ -37,38 +39,25 @@ class CityMainActivity : AppCompatActivity(), CityContract.View {
...
-
-    private fun createCityList(json: String?) {
-        val jsonArray = JSONArray(json)
-        setCityListAdapter(viewModel.createCityList(jsonArray))
-    }
-
-    private fun setCityListAdapter(cities: MutableList<String>) {
-        val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, cities)
+    private fun setCityListAdapter() {
+        val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, JSONDataUseCase.getJSONData())
```
Comment (SartiMau) [inline WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt]: the useCase should be called from the ViewModel not the activity

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt (hunk 8 above)
Comment (BQuinteros) [reply WheaterApp/app/src/main/java/com/example/weatherapp/activities/CityMainActivity.kt]: Done

**Review by SartiMau — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]

**Pull request merged**